src/storage/aggregator.cc

```diff
@@ -0,0 +1,383 @@
+        }
+        default: {
+            PDLOG(ERROR, "Unsupported timestamp data type");
+            return false;
+        }
+    }
+
+    AggrBufferLocked* aggr_buffer_lock;
+    {
+        std::lock_guard<std::mutex> lock(mu_);
+        if (aggr_buffer_map_.find(key) == aggr_buffer_map_.end()) {
+            aggr_buffer_map_.emplace(key, AggrBufferLocked{});
+        }
+        aggr_buffer_lock = &aggr_buffer_map_.at(key);
+    }
+
+    std::unique_lock<std::mutex> lock(*aggr_buffer_lock->mu_);
+    AggrBuffer& aggr_buffer = aggr_buffer_lock->buffer_;
+
+    // init buffer timestamp range
+    if (aggr_buffer.ts_begin_ == -1) {
+        aggr_buffer.ts_begin_ = cur_ts;
+        if (window_type_ == WindowType::kRowsRange) {
+            aggr_buffer.ts_end_ = cur_ts + window_size_ - 1;
+        }
+    }
+
+    if (CheckBufferFilled(cur_ts, aggr_buffer.ts_end_, aggr_buffer.aggr_cnt_)) {
+        AggrBuffer flush_buffer = aggr_buffer;
+        int64_t latest_ts = aggr_buffer.ts_end_ + 1;
+        aggr_buffer.clear();
+        aggr_buffer.ts_begin_ = latest_ts;
+        if (window_type_ == WindowType::kRowsRange) {
+            aggr_buffer.ts_end_ = latest_ts + window_size_ - 1;
+        }
+        lock.unlock();
+        FlushAggrBuffer(key, flush_buffer);
+        lock.lock();
+    }
+
+    if (cur_ts < aggr_buffer.ts_begin_) {
+        // handle the case that the current timestamp is smaller than the begin timestamp in aggregate buffer
+        lock.unlock();
+        bool ok = UpdateFlushedBuffer(key, row_ptr, cur_ts, offset);
+        if (!ok) {
+            PDLOG(ERROR, "Update flushed buffer failed");
+            return false;
+        }
+    } else {
+        aggr_buffer.aggr_cnt_++;
+        aggr_buffer.binlog_offset_ = offset;
+        if (window_type_ == WindowType::kRowsNum) {
+            aggr_buffer.ts_end_ = cur_ts;
+        }
+        bool ok = UpdateAggrVal(base_row_view_, row_ptr, &aggr_buffer);
+        if (!ok) {
+            PDLOG(ERROR, "Update aggr value failed");
+            return false;
+        }
+    }
+    return true;
+}
+
+bool Aggregator::GetAggrBuffer(const std::string& key, AggrBuffer* buffer) {
+    std::lock_guard<std::mutex> lock(mu_);
+    if (aggr_buffer_map_.find(key) == aggr_buffer_map_.end()) {
+        return false;
+    }
+    *buffer = aggr_buffer_map_.at(key).buffer_;
+    return true;
+}
+
+bool Aggregator::GetAggrBufferFromRowView(const codec::RowView& row_view, const int8_t* row_ptr, AggrBuffer* buffer) {
+    if (buffer == nullptr) {
+        return false;
+    }
+    row_view.GetValue(row_ptr, 1, DataType::kTimestamp, &buffer->ts_begin_);
+    row_view.GetValue(row_ptr, 2, DataType::kTimestamp, &buffer->ts_end_);
+    row_view.GetValue(row_ptr, 3, DataType::kInt, &buffer->aggr_cnt_);
+    char* ch = NULL;
+    uint32_t ch_length = 0;
+    row_view.GetValue(row_ptr, 4, &ch, &ch_length);
+    switch (aggr_col_type_) {
+        case DataType::kSmallInt:
+        case DataType::kInt:
```

Comment:
why not use `vint` and `vsmallint`

Reply:
for possible overflow of `aggr_val` in `SumAggregator`